google/cloud/spanner/internal/session_pool_test.cc

```diff
@@ -0,0 +1,126 @@
+#include "google/cloud/status.h"
+#include "google/cloud/testing_util/assert_ok.h"
+#include <gmock/gmock.h>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace internal {
+namespace {
+
+using ::testing::ByMove;
+using ::testing::Return;
+
+class MockSessionManager : public SessionManager {
+ public:
+  MOCK_METHOD1(CreateSessions,
+               StatusOr<std::vector<std::unique_ptr<Session>>>(size_t));
+};
+
+// Helper to build a vector of Sessions.
+std::vector<std::unique_ptr<Session>> MakeSessions(
+    std::vector<std::string> session_names) {
+  std::vector<std::unique_ptr<Session>> ret;
+  ret.reserve(session_names.size());
+  for (auto& name : session_names) {
+    ret.push_back(
+        google::cloud::internal::make_unique<Session>(std::move(name)));
+  }
+  return ret;
+}
+
+TEST(SessionPool, Allocate) {
+  auto mock = google::cloud::internal::make_unique<MockSessionManager>();
+  EXPECT_CALL(*mock, CreateSessions(1))
+      .WillOnce(Return(ByMove(MakeSessions({"session1"}))));
+
+  SessionPool pool(mock.get());
+  auto session = pool.Allocate();
+  ASSERT_STATUS_OK(session);
+  EXPECT_EQ((*session)->session_name(), "session1");
+}
+
+TEST(SessionPool, CreateError) {
+  auto mock = google::cloud::internal::make_unique<MockSessionManager>();
+  EXPECT_CALL(*mock, CreateSessions(1))
+      .WillOnce(Return(ByMove(Status(StatusCode::kInternal, "some failure"))));
+
+  SessionPool pool(mock.get());
+  auto session = pool.Allocate();
+  EXPECT_EQ(session.status().code(), StatusCode::kInternal);
+  EXPECT_EQ(session.status().message(), "some failure");
+}
+
+TEST(SessionPool, ReuseSession) {
+  auto mock = google::cloud::internal::make_unique<MockSessionManager>();
+  EXPECT_CALL(*mock, CreateSessions(1))
+      .WillOnce(Return(ByMove(MakeSessions({"session1"}))));
+
+  SessionPool pool(mock.get());
+  auto session = pool.Allocate();
+  ASSERT_STATUS_OK(session);
+  EXPECT_EQ((*session)->session_name(), "session1");
+  pool.Release((*session).release());
+
+  auto session2 = pool.Allocate();
+  ASSERT_STATUS_OK(session2);
+  EXPECT_EQ((*session2)->session_name(), "session1");
+  pool.Release((*session2).release());
+}
+
+TEST(SessionPool, Lifo) {
+  auto mock = google::cloud::internal::make_unique<MockSessionManager>();
+  EXPECT_CALL(*mock, CreateSessions(1))
+      .WillOnce(Return(ByMove(MakeSessions({"session1"}))))
+      .WillOnce(Return(ByMove(MakeSessions({"session2"}))));
+
+  SessionPool pool(mock.get());
+  auto session = pool.Allocate();
+  ASSERT_STATUS_OK(session);
+  EXPECT_EQ((*session)->session_name(), "session1");
+
+  auto session2 = pool.Allocate();
+  ASSERT_STATUS_OK(session2);
+  EXPECT_EQ((*session2)->session_name(), "session2");
+
+  pool.Release((*session).release());
+  pool.Release((*session2).release());
+
+  // We should get the sessions back in the reverse order they were released.
```

Comment:
Perhaps expound on why this is (currently?) important.

Reply:
> Done.

OK, although I'd still say it is unclear why this is a necessary/desirable property.